### PR TITLE
 CLOUD-62 [PXC] PodDisruptionBudget should be created by default

### PR DESCRIPTION
**Pull request by delgod**

### CODE REVIEW

File: pkg/pxc/statefulset.go [hunk 1]
```diff
@@ -132,3 +134,21 @@ func getConfigVolumes(cr *api.PerconaXtraDBCluster, cvName string) corev1.Volume
...
+	trueVar := true
+
+	return metav1.OwnerReference{
+		APIVersion: gvk.GroupVersion().String(),
+		Kind:       gvk.Kind,
+		Name:       sfs.GetName(),
+		UID:        sfs.GetUID(),
+		Controller: &trueVar,
+	}, nil
+}
```
Comment (dAdAbird) [inline pkg/pxc/statefulset.go]: Why do we need this function? I don't see any references to it.

File: pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go [hunk 2]
```diff
@@ -364,11 +399,30 @@ func (r *ReconcilePerconaXtraDBCluster) deletePVC(namespace string, lbls map[str
...
+// OwnerRef returns OwnerReference to object
+func OwnerRef(ro runtime.Object, scheme *runtime.Scheme) (metav1.OwnerReference, error) {
+	gvk, err := apiutil.GVKForObject(ro, scheme)
+	if err != nil {
+		return metav1.OwnerReference{}, err
+	}
+
+	trueVar := true
+
+	ca, err := meta.Accessor(ro)
```
Comment (dAdAbird) [inline pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go]: Should be an error handling.

File: pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go [hunk 3]
```diff
@@ -264,6 +283,22 @@ func (r *ReconcilePerconaXtraDBCluster) deploy(cr *api.PerconaXtraDBCluster) err
...
+			err = setControllerReference(proxySet, pdbProxySQL, r.scheme)
+			if err != nil {
+				return err
+			}
+
+			err = r.client.Create(context.TODO(), pdbProxySQL)
+			if err != nil && !errors.IsAlreadyExists(err) {
+				return fmt.Errorf("create PodDisruptionBudget-PXC: %v", err)
+			}
+		}
```
Comment (dAdAbird) [inline pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go]: Should be an appropriate error handling. What if `err != nil`?

File: pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go [hunk 4]
```diff
@@ -239,8 +241,25 @@ func (r *ReconcilePerconaXtraDBCluster) deploy(cr *api.PerconaXtraDBCluster) err
...
+		err = setControllerReference(nodeSet, pdbPXC, r.scheme)
+		if err != nil {
+			return err
+		}
+
+		err = r.client.Create(context.TODO(), pdbPXC)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			return fmt.Errorf("create PodDisruptionBudget-PXC: %v", err)
+		}
+	}
```
Comment (dAdAbird) [inline pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go]: Should be an appropriate error handling. What if `err != nil`?

File: pkg/pxc/statefulset.go (hunk 1 above)
Comment (Romanches) [reply pkg/pxc/statefulset.go]: Do you mean StatefulSetOwnerRef() from statefulset.go? 
Hm. It is not mine. Should I delete it?

File: pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go (hunk 3 above)
Comment (Romanches) [reply pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go]: In this keys only two ways are posible:
1) if errors.IsAlreadyExists(err) -> Do nothing
2) if !errors.IsAlreadyExists(err) -> return err

I guess this part works correct. Isn't it?

File: pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go (hunk 2 above)
Comment (Romanches) [reply pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go]: Agree.

Comment (Romanches) [reply pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go]: Done!

File: pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go (hunk 3 above)
Comment (dAdAbird) [reply pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go]: meant not lines 298, but error handling in the block L289-301:
starting here https://github.com/Percona-Lab/percona-xtradb-cluster-operator/pull/85/files/cc25b8ecd6d9bb4a91e7bd26778b7daf6452d1ab#diff-f446014a9240c7ae115b90dc2fa50e6fR289 
if 
```go
err = r.client.Get(context.TODO(), types.NamespacedName{Name: proxySet.Name, Namespace: proxySet.Namespace}, proxySet)
```
in L289 returns an error the Distributed budget creation will be skiped without any notices

File: pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go [hunk 5]
```diff
@@ -264,6 +286,26 @@ func (r *ReconcilePerconaXtraDBCluster) deploy(cr *api.PerconaXtraDBCluster) err
 		if err != nil && !errors.IsAlreadyExists(err) {
 			return fmt.Errorf("create PXC Service: %v", err)
 		}
+
+		// PodDistributedBudget object for ProxySQL
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: proxySet.Name, Namespace: proxySet.Namespace}, proxySet)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return fmt.Errorf("Get stateful set %s: %v", proxySet.Name, err)
```
Comment (dAdAbird) [inline pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go]: Error strings should not be capitalized: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Comment (dAdAbird) [reply pkg/controller/perconaxtradbcluster/perconaxtradbcluster_controller.go]: and I'm not sure that `proxySet.Name` is safe to use here. You sure in can't a nil after Get returns with error? Just put "get proxy stateful set: %v"
\* _the same above_

**Pull request merged**